_posts/2017-07-12-TypeScript-for-JS-devs.markdown

```diff
@@ -0,0 +1,485 @@
+
+Interfaces are a great way to define contracts within you code. Interfaces provide shape when declaring things like variables. 
+
+The person object fails to compile because an object with type of person expects to have the property isMillennial.
+
+This means when you refactor, then you’ll start getting messages throughout your app saying where changes need to be applied. This means you can trust your code a lot more.
+
+```ts
+interface Person {
+  name: string
+  interest: any
+  isMillennial: boolean
+  phoneNumber?: string
+}
+
+const myPerson: Person = {
+  name: "Jane",
+  interest: ["music", "reading", "dancing"],
+  phoneNumber: "555-555-5555"
+}
+```
+
+A very common technique in our code is to extend interfaces:
+
+```ts
+interface ArtsyPerson extends Person {
+  role: string
+  joined: string
+  floor?: string
+}
+
+const artsyEmployee: ArtsyPerson = {
+  name: "Christina",
+  role: "Engineer",
+  joined: "2 years ago",
+  floor: "25"
+}
+
+```
+
+You might wonder how this fits with ES6 classes? You can use classes as at Type. TypeScript classes are not much different then the first class language construct in Traditional JS. In TypeScript you can use classes to define a specific type.
+
+```ts
+interface Point {
+  x: number 
+  y: number
+}
+
+class MyPoint implements Point {
+   x: number; y: number; 
+}
+
+let point: MyPoint = new MyPoint()
+point.x = 3;
+point.y = 8;
+
+point.y = "10"; // compile error
+```
+
+# Generics
+
+
+
+TypeScript (especially code we write with React) heavily uses generics to handle related classes. 
+
+In this case, we’re saying we want an Array with an associated type of string. This allows TypeScript to know what type of object would be returned from a get function.
+
+```ts
+const names: Array<string> = [
+  "christina", 
+  "orta",
+  "danger"
+]
+```
+
+Without generics, you’d have to create a version of the array that only returns a specific type. Instead you would say “this is an array - and everything inside it is going to be a string” [...]
+
+```ts
+const names: ArrayString = [
+  "christina", 
+  "orta",
+  "danger"
+]
+```
+
+There is shorthand for this in the form of `array[]`.
+
+## Generic Promises
+
+Let’s look at another example - Promises
+
+In this case we are stating that the promise will return a number. This associates the returned value inside the promise with the API. It means we would get compiler errors if that is treated as a string.
+
+```ts
+const getAge:Promise<number> = new Promise(res => 5)
+
+const age = await getAge()
+if (age === "18") {
+  ...
+}
+```
+
+## Generic Functions
+
+These are useful for when you have a function which allows you to work with many types, and keep them consistent. 
+
+For example this reverse array function will take a type called T, which comes from the param and it will return an array of the same type
+
+This is generally only useful for utilities funds, classes or libraries
+
+```ts
+function reverseArray<T>(arr:T[]): T[] {
```

Comment:
As this signature is somewhat dense, it might be good to expand a bit on what `T` is -- is it a placeholder type? a variable (could `T` be replaced with `Y` for example)? Is this a feature specifically defined by TypeScript, or does it follow a larger pattern?

Reply:
Yes, `T` is a variable and could be replaced with `Y`. `T` is the most commonly used name, presumably it stands for “type”, but in e.g. React you’ll have `P` for props and `S` for state.

Reply:
Oh right, in the blog post :) We should assume that the reader has no previous knowledge of types / comes from a JS background.